pkg/ebpf/events_enrich.go

```diff
@@ -8,172 +8,163 @@ import (
-					delete(attempted, cgroupId)
-					attemptedMutex.Unlock()
-				}
-				// non container event and not cgroup_mkdir (or the event already enriched) skip per cgroupId caching
-				if (event.ContainerID == "" && event.EventID != int(events.CgroupMkdir)) || event.ContainerImage != "" {
-					// enrichment attempt is being done, or has failed, so do not try it again
-					attemptedMutex.RLock()
-					if _, ok := attempted[cgroupId]; ok {
-						// wait transaction (cache flush) to finish so cached events and this one are ordered
-						attempted[cgroupId].Lock()
-						out <- event
-						attempted[cgroupId].Unlock()
-						attemptedMutex.RUnlock()
-						continue
+					bLock.RLock()
+					if i, ok := enrichInfo[cgroupId]; ok {
+						eInfo = i
 					}
-					attemptedMutex.RUnlock()
-					out <- event // some events might not have an attempted map
-					continue
-				} else {
-					//if the container queue has not been created - create the container queue and invoke to enrich query
-					queuesMutex.RLock()
-					_, exists := queues[cgroupId]
-					queuesMutex.RUnlock()
-					if !exists {
-						queuesMutex.Lock()
-						queues[cgroupId] = make(chan *trace.Event, enrichmentQueueSize)
-						queuesMutex.Unlock()
-
-						go func() {
-							metadata, err := t.containers.EnrichCgroupInfo(cgroupId)
-							enriches <- enrichResult{cgroupId, metadata, err}
-						}()
-					}
-					queuesMutex.RLock()
-					queue, ok := queues[cgroupId]
-					queuesMutex.RUnlock()
-					if ok {
-						//if queue still exists send as usual
-						queue <- event
-					} else {
-						//if not, get the enriched data and handle the stale queue
-						info := t.containers.GetCgroupInfo(cgroupId)
-						queuesMutex.RLock()
-						queue = staleQueues[cgroupId]
-						queuesMutex.RUnlock()
-
-						//probably empty but just to make sure
-						for evt := range queue {
-							enrichEvent(evt, info.Container)
-							out <- evt
+					for evt := range queues[cgroupId] {
+						if evt.ContainerImage == "" && eInfo != nil {
+							bLock.RUnlock()
+							enrichEvent(evt, eInfo.result)
+							bLock.RLock()
 						}
-
-						enrichEvent(event, info.Container)
-
-						close(queue)
-						queuesMutex.Lock()
-						delete(staleQueues, cgroupId)
-						queuesMutex.Unlock()
-						out <- event
+						out <- evt
 					}
+					bLock.RUnlock()
+					bLock.Lock()
+					delete(enrichDone, cgroupId)
+					delete(enrichInfo, cgroupId)
+					delete(queues, cgroupId)
+					bLock.Unlock()
+					continue
+				}
+				// make sure a queue channel exists for this cgroupId
+				bLock.Lock()
+				if _, ok := queues[cgroupId]; !ok {
+					queues[cgroupId] = make(chan *trace.Event, contQueueSize)
+					enrichDone[cgroupId] = &uberAtom.Bool{}
+
+					go func(cgroupId uint64) {
+						metadata, err := t.containers.EnrichCgroupInfo(cgroupId)
+						bLock.Lock()
+						enrichInfo[cgroupId] = &enrichResult{metadata, err}
+						enrichDone[cgroupId].Store(true)
+						bLock.Unlock()
```

Comment:
I'm pretty sure there's a race condition that can happen between this line 132. And this race condition would panic tracee if i'm right that it can happen.
I think you should just lock 118-119 and only read lock the rest.

Reply:
If I understand correctly the semantics of this rwmutex,
This is a different go routine.
So from what I understand, this code won't run until the current go routine will release it's lock in 132.

Reply:
You're right just went over this with Rafael, it will probably be fine. I'd still fine grain that lock just to be sure, but we'll probably be fine without.

Reply:
Yep, its a diff routine and it can Lock/Unlock. I could Unlock()/Lock() before 130 but we're fine (same numbers from previous test).